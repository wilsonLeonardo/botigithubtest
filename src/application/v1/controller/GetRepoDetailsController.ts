import { RequestHandler } from 'express';
import { OK } from 'http-status';
import { ILogger } from '@infrastructure/logger/ILogger';
import { buildErrorInfo } from '@infrastructure/parser/ErrorInfo';
import IGitHubService from '@service/IGitHubService';

export default class GetRepoDetailsController {
  constructor(private gitHubService: IGitHubService, private logger: ILogger) {}

  public execute: RequestHandler = async (request, response) => {
    try {
      const { repoId } = request.params;

      const repository = await this.gitHubService.getRepoDetails(Number(repoId));

      return response.status(OK).json(repository);
    } catch (err: any) {
      const errorInfo = buildErrorInfo(err);

      this.logger.error(`fail to get repository details`);

      return response.status(errorInfo.code).json({ error: errorInfo.errorMessage });
    }
  };
}
