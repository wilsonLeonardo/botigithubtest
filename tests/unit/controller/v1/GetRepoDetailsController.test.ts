import { Request } from 'express';
import { INTERNAL_SERVER_ERROR, OK } from 'http-status';
import GetRepoDetailsController from '../../../../src/application/v1/controller/GetRepoDetailsController';
import GitHubServiceMock from '../../../mocks/GitHubServiceMock';
import LoggerMock from '../../../mocks/LoggerMock';
import InternalServerError from '../../../../src/domain/exceptions/InternalServerError';

describe('GetRepoDetailsController', () => {
  const githubService = new GitHubServiceMock();
  const logger = new LoggerMock();
  it('should match the snapshot #unit', () => {
    const instance = new GetRepoDetailsController(githubService, logger);
    expect(instance).toMatchSnapshot();
  });
  it('should get repository details #unit', async () => {
    const detail = { xpto: 'xpto' };
    githubService.getRepoDetails = jest.fn().mockImplementation((): Promise<any> => {
      return Promise.resolve(detail);
    });
    const instance = new GetRepoDetailsController(githubService, logger);

    const req = { params: { repoId: 242 } } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(OK);
    expect(mockJsonResponse).toHaveBeenCalledWith(detail);
  });
  it('should return error code when service throws #unit', async () => {
    githubService.getRepoDetails = jest.fn().mockImplementation((): Promise<any> => {
      return Promise.reject(new InternalServerError('error message'));
    });
    const instance = new GetRepoDetailsController(githubService, logger);

    const req = { params: { repoId: 242 } } as unknown as Request;
    const mockJsonResponse = jest.fn();
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: mockJsonResponse,
      })),
    };

    await instance.execute(req, res, jest.fn());
    expect(res.status).toHaveBeenCalledWith(INTERNAL_SERVER_ERROR);
    expect(mockJsonResponse).toHaveBeenCalledWith({ error: 'error message' });
  });
});
