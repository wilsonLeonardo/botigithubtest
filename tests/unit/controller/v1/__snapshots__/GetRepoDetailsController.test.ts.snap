// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetRepoDetailsController should match the snapshot #unit 1`] = `
GetRepoDetailsController {
  "execute": [Function],
  "gitHubService": GitHubServiceMock {},
  "logger": LoggerMock {},
}
`;
